### PR TITLE
fix: handle v2 models in func signatures

### LINKED ISSUE #10965: Using `BaseModel` default args in flows raises JSON-related TypeError
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

Deploying flows with `BaseModel` default args raises a JSON-related `TypeError`. I expected this to work as of https://github.com/PrefectHQ/prefect/pull/10946 (`prefect==2.13.7`).

This works with `pydantic<2`, so it might be an upstream issue. I don't think so though, since plain `dumps` on pydantic objects will fail regardless of `pydantic` version.

### Reproduction

```python3
from prefect import flow
from pydantic import BaseModel

class FlowArg(BaseModel):
    value: str

# This fails regardless of pydantic version
# from json import dumps
# dumps(FlowArg(value="whatever")

# this fails only for pydantic>2
@flow
def flow(arg: FlowArg = FlowArg(value="sane-default")):
    print(f"{arg.value=}")

if __name__ == "__main__":
    flow()
```

### Error

```python3
Traceback (most recent call last):
  File "mre.py", line 9, in <module>
    @flow
     ^^^^
  File "<snip>/.venv/lib/python3.11/site-packages/prefect/flows.py", line 1133, in flow
    Flow(
  File "<snip>/.venv/lib/python3.11/site-packages/prefect/context.py", line 185, in __register_init__
    __init__(__self__, *args, **kwargs)
  File "<snip>/.venv/lib/python3.11/site-packages/prefect/flows.py", line 292, in __init__
    self.parameters = parameter_schema(self.fn)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<snip>/.venv/lib/python3.11/site-packages/prefect/utilities/callables.py", line 308, in parameter_schema
    ).schema(by_alias=True)
      ^^^^^^^^^^^^^^^^^^^^^
  File "<snip>/.venv/lib/python3.11/site-packages/pydantic/v1/main.py", line 664, in schema
    s = model_schema(cls, by_alias=by_alias, ref_template=ref_template)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<snip>/.venv/lib/python3.11/site-packages/pydantic/v1/schema.py", line 188, in model_schema
    m_schema, m_definitions, nested_models = model_process_schema(
                                             ^^^^^^^^^^^^^^^^^^^^^
  File "<snip>/.venv/lib/python3.11/site-packages/pydantic/v1/schema.py", line 582, in model_process_schema
    m_schema, m_definitions, nested_models = model_type_schema(
                                             ^^^^^^^^^^^^^^^^^^
  File "<snip>/.venv/lib/python3.11/site-packages/pydantic/v1/schema.py", line 623, in model_type_schema
    f_schema, f_definitions, f_nested_models = field_schema(
                                               ^^^^^^^^^^^^^
  File "<snip>/.venv/lib/python3.11/site-packages/pydantic/v1/schema.py", line 249, in field_schema
    s, schema_overrides = get_field_info_schema(field)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<snip>/.venv/lib/python3.11/site-packages/pydantic/v1/schema.py", line 217, in get_field_info_schema
    schema_['default'] = encode_default(field.default)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<snip>/.venv/lib/python3.11/site-packages/pydantic/v1/schema.py", line 996, in encode_default
    return pydantic_encoder(dft)
           ^^^^^^^^^^^^^^^^^^^^^
  File "<snip>/.venv/lib/python3.11/site-packages/pydantic/v1/json.py", line 90, in pydantic_encoder
    raise TypeError(f"Object of type '{obj.__class__.__name__}' is not JSON serializable")
TypeError: Object of type 'FlowArg' is not JSON serializable
```

### Versions

```Text
`pydantic` version: 2.4.2

Version:             2.13.7
API version:         0.8.4
Python version:      3.11.5
Git commit:          70f2d40d
Built:               Fri, Oct 13, 2023 8:20 AM
OS/Arch:             linux/x86_64
Profile:             local
Server type:         server
```

### Additional context

This works with `pydantic<2`, so might be an upstream issue. The trace is littered with `/pydantic/v1` which seems strange.

### PR DESCRIPTION
**Pull request by urimandujano**
Closes #10965

This is an addition to our pydantic v2 parameter schema generation logic. When we parse a flow function we attempt to generate a schema for it. Our schema generation logic is entirely pydantic v1 based. Now that we've lifted our restriction on pydantic v2, it's possible for flow function signatures to make use of pydantic v2-based models. The error was coming from using v1-schema generation on a v2 model. 

~Note that with this fix the schema that is generated will drop the default with a warning:~ This is no longer an issue with the new implementation!

### Checklist

- [X] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [X] This pull request includes tests or only affects documentation.
- [X] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`, `docs`.

### CODE REVIEW

File: tests/utilities/test_callables.py [hunk 1]
```diff
@@ -297,6 +297,22 @@ def f(x: Foo):
...
+        """
+        import pydantic
+
+        class Foo(pydantic.BaseModel):
+            bar: str
+
+        def f(foo: Foo = Foo(bar="baz")):
+            ...
+
+        callables.parameter_schema(f)
```
Comment (zzstoatzz) [inline tests/utilities/test_callables.py:314]: is the resulting schema the same in both cases?

Comment (urimandujano) [reply tests/utilities/test_callables.py:314]: It looks like a couple of things changed in pydantic v2 -- one is that the field `definitions` gets replaced with the standard `$defs` which we end up dropping: https://www.diffchecker.com/HuP7QiSk/

The other is that if we copy `$defs` into `definitions`, v2 generates a slightly differently formatted schema: https://www.diffchecker.com/pFFlfjDW/
 
I'm copying the `$defs` into `definitions` but there are a bunch of differences in the generated schemas.

**Review by zzstoatzz — APPROVED** [review]
this makes sense to me.

I'd also be curious for @desertaxle's thoughts on this, just wondering about whether there could be certain cases where the schema ends up different and mess with the UI forms for flow run params

**Comment by desertaxle** [comment]
> I'd also be curious for @desertaxle's thoughts on this, just wondering about whether there could be certain cases where the schema ends up different and mess with the UI forms for flow run params

Are there examples of what the schema would look like if `pydantic` v2 is used? As long as the top-level keys are consistent, I think the UI should be able to handle it, and it would be worthwhile to capture that in a test to avoid surprises.

File: tests/utilities/test_callables.py [hunk 2]
```diff
@@ -297,6 +301,131 @@ def f(x: Foo):
...
+        }
+        duration_schema = {
+            "title": "pduration",
+            "default": 5.0,
+            "position": 8,
+            "type": "number",
+            "format": "time-delta",
+        }
+
+        if HAS_PYDANTIC_V2:
```
Comment (urimandujano) [inline tests/utilities/test_callables.py:375]: These are the only remaining JsonSchema changes between pydantic v1 and v2. I think the reason is that v2 using a [standardized version of JsonSchema now](https://docs.pydantic.dev/latest/concepts/json_schema/).

**Comment by urimandujano** [comment]
This was a bit trickier than I originally thought because pydantic v2 can't generate JSONSchema for `pendulum` data types or user defined classes. Even after smoothing over the differences in schemas for built-in types (those changes are contained in `create_v2_schemas`, those two conditions ^ made the v2 schemas incompatible with our v1 schemas.
 
 This PR is doing a few things now:
- adds pydantic v1 and v2 support for function JSONSchema generation
- smooths over differences in v1 and v2 schemas
- extends pydantic's v2 JSONSchema for pendulum's `DateTime`, `Date`, and `Duration` types
- extends pydantic's v2 JSONSchema for user-defined types
- detects if any function parameters are v2 and if so, generates a v2 JSONSchema 
  - this last point is important because it should fix issues in our integrations [such as this one](https://github.com/PrefectHQ/prefect-aws/issues/325). In the case where pydantic v2 is installed, our integrations use pydantic's vendored v1. Previously we would detect v2 was installed and expect v2 models. Now we can handle that case where we are working with a v1 model with v2 installed.  
  
Sorry that this PR became such a moshpit - it was impossible to get the tests passing without fixing all of these issues at once.

File: src/prefect/cli/deployment.py [hunk 3]
```diff
@@ -601,6 +603,7 @@ async def run(
         )
 
         try:
+            print(Scheduled(scheduled_time=scheduled_start_time).state_details)
```
Comment (chrisguidry) [inline src/prefect/cli/deployment.py:606]: Stray `print`?

File: src/prefect/cli/deployment.py [hunk 4]
```diff
@@ -621,6 +624,7 @@ async def run(
         run_url = "<no dashboard available>"
 
     datetime_local_tz = scheduled_start_time.in_tz(pendulum.tz.local_timezone())
+    print(datetime_local_tz)
```
Comment (chrisguidry) [inline src/prefect/cli/deployment.py:627]: Stray?

File: tests/utilities/test_callables.py [hunk 5]
```diff
@@ -297,6 +301,131 @@ def f(x: Foo):
...
+            # these overrides represent changes in how pydantic generates schemas in v2
+            datetime_schema["default"] = "2025-01-01T00:00:00Z"
+            duration_schema["default"] = "PT5S"
+            duration_schema["type"] = "string"
+            duration_schema["format"] = "duration"
+
+        schema = callables.parameter_schema(f)
+        print(schema.dict())
```
Comment (chrisguidry) [inline tests/utilities/test_callables.py:383]: Another stray

**Review by chrisguidry — APPROVED** [review]
Okay, this all makes sense to me, and thank you for the new test that confirms what the schema should look like!

I was wondering, so I double-checked that our CI unit tests are doing the right thing.  They use the `scripts/generate-lower-bounds.py` to produce the requirements file for most of the test suite, and the lower-bounds still produces `pydantic==1.10`.  Our other pydantic v2 tests manually install the latest `pydantic>2`, so we should be getting coverage under both versions.

**Comment by urimandujano** [comment]
Thank you for the catch @chrisguidry, those stray prints should be gone now. To your point about testing two versions, I figured out that re-importing pydantic within a test function will pull in pydantic v2 so I've been abusing that detail 😓 

@aaazzam I did see some differences but they seem mostly confined to metadata differences (v1 doesn't have a `type` key whereas v2 doesn't have a `description` key. Check out this [commit](https://github.com/PrefectHQ/prefect/pull/10966/commits/5ef288fbcc52902a1c139dcaabfbeddb8b36b13f) to see what we have to do to make them compat.

**Pull request merged**